app/scripts/modules/google/src/instance/custom/customInstanceBuilder.gce.service.js

```diff
@@ -10,57 +10,100 @@ export const name = GOOGLE_INSTANCE_CUSTOM_CUSTOMINSTANCEBUILDER_GCE_SERVICE; //
         return true;
       }
-      return vCpuCount % 2 === 0;
+      switch (instanceFamily) {
+        case 'N2':
+        case 'N2D':
+          return vCpuCount % 4 === 0;
+        case 'E2':
+        case 'N1':
+        default:
+          return vCpuCount % 2 === 0;
+      }
     }
 
     function generateValidVCpuListForLocation(location, locationToInstanceTypesMap) {
       const max = locationToInstanceTypesMap[location].vCpuMax;
       return [1, ..._.range(2, max, 2), max];
     }
 
+    function generateValidInstanceFamilyList() {
+      return ['n1', 'e2', 'n2', 'n2d'].map(x => x.toUpperCase());
+    }
+
     /*
      * Memory per vCPU must be between .9 GB and 6.5 GB
      * Total memory must be a multiple of 256 MB.
      * */
-    function minMemoryForVCpuCount(vCpuCount) {
-      return Math.ceil(0.9 * vCpuCount * 4) / 4;
+    function minMemoryForVCpuCount(instanceFamily, vCpuCount) {
+      switch (instanceFamily) {
+        case 'E2':
+        case 'N2':
+        case 'N2D':
+          return Math.ceil(0.5 * vCpuCount * 4) / 4;
+        case 'N1':
+        default:
+          return Math.ceil(0.9 * vCpuCount * 4) / 4;
+      }
     }
 
-    function maxMemoryForVCpuCount(vCpuCount) {
-      return 6.5 * vCpuCount;
+    function maxMemoryForVCpuCount(instanceFamily, vCpuCount) {
+      switch (instanceFamily) {
+        case 'E2':
+          return 8 * vCpuCount > 128 ? 128 : 8 * vCpuCount;
+        case 'N2':
+        case 'N2D':
+          return 8 * vCpuCount;
+        case 'N1':
+        default:
+          return 6.5 * vCpuCount;
+      }
     }
 
-    function generateValidMemoryListForVCpuCount(vCpuCount) {
-      const min = minMemoryForVCpuCount(vCpuCount);
-      const max = maxMemoryForVCpuCount(vCpuCount);
+    function generateValidMemoryListForVCpuCount(instanceFamily, vCpuCount) {
+      const min = minMemoryForVCpuCount(instanceFamily, vCpuCount);
+      const max = maxMemoryForVCpuCount(instanceFamily, vCpuCount);
       return [..._.range(min, max, 0.25), max];
     }
 
-    function memoryIsValid(totalMemory, vCpuCount) {
-      const min = minMemoryForVCpuCount(vCpuCount);
-      const max = maxMemoryForVCpuCount(vCpuCount);
+    function memoryIsValid(instanceFamily, totalMemory, vCpuCount) {
+      const min = minMemoryForVCpuCount(instanceFamily, vCpuCount);
+      const max = maxMemoryForVCpuCount(instanceFamily, vCpuCount);
       return _.inRange(totalMemory, min, max) || totalMemory === max;
     }
 
     /*
      * In the API, you must always provide memory in MB units.
      * Format: custom-NUMBER_OF_CPUS-AMOUNT_OF_MEMORY
      * */
-    function generateInstanceTypeString(vCpuCount, totalMemory) {
+    function generateInstanceTypeString(instanceFamily, vCpuCount, totalMemory) {
       const memoryInMbs = Number(totalMemory) * 1024;
-      return `custom-${vCpuCount}-${memoryInMbs}`;
+      instanceFamily = instanceFamily.toLowerCase();
+      if (instanceFamily === 'n1') {
```

Comment:
Just curious why we are special-casing the `n1` type and displaying `custom` instead of `n1`? Is it because `n1` is already well-known to be a custom type and showing `n1-custom-...` would be redundant?

Reply:
Today `custom-xx-xxxx` it's N1 type by default and this it's good for backward compatibility